include/mma_tensor_op/mma_m16n16k16_sm70.cuh

```diff
@@ -0,0 +1,216 @@
+      const int idx_contiguous = warp_row * 8 + wrm.quad_row * 4 + wrm.quad_hilo * 2;
+      const int thread_offset_a = idx_strided * stride + idx_contiguous;
+      reg[0] = A[thread_offset_a + 0];
+      reg[1] = A[thread_offset_a + 1];
+    }
+  };
+
+  template <int stride> struct MmaOperandB {
+
+    unsigned reg[2];
+
+    __device__ inline void load(void *smem, int k, int warp_col, const WarpRegisterMapping &wrm)
+    {
+      unsigned *B = reinterpret_cast<unsigned *>(smem);
+      const int idx_strided = k * 4 + wrm.quad_thread;
+      const int idx_contiguous = warp_col * 8 + wrm.quad_col * 4 + wrm.quad_hilo * 2;
+      const int thread_offset_b = idx_strided * stride + idx_contiguous;
+      reg[0] = B[thread_offset_b + 0];
+      reg[1] = B[thread_offset_b + 1];
+    }
+  };
+
+  template <int stride, class store_type> struct MmaOperandC {
+  };
+
+  template <int stride> struct MmaOperandC<stride, half> {
+
+    using reg_type = unsigned;
+    reg_type reg[4];
+
+    __device__ MmaOperandC()
+    {
+#pragma unroll
+      for (int i = 0; i < 4; i++) { reg[i] = 0; }
+    }
+
+    __device__ void store(void *smem, int warp_row, int warp_col, const WarpRegisterMapping &wrm)
+    {
+      unsigned *C = reinterpret_cast<unsigned *>(smem);
+
+      const int idx_strided = warp_row * 16 + wrm.quad_row * 8 + wrm.quad_hilo * 4 + wrm.quad_thread;
+      const int idx_contiguous = warp_col * 8 + wrm.quad_col * 4;
+      const int thread_offset_c = idx_strided * stride + idx_contiguous;
+#pragma unroll
+      for (int i = 0; i < 4; i++) { C[thread_offset_c + i] = reg[i]; }
+    }
+  };
+
+  template <int stride> struct MmaOperandC<stride, float> {
+
+    using reg_type = float;
+    reg_type reg[8];
+
+    __device__ MmaOperandC()
+    {
+#pragma unroll
+      for (int i = 0; i < 8; i++) { reg[i] = 0; }
+    }
+
+    __device__ void store(void *smem, int warp_row, int warp_col, const WarpRegisterMapping &wrm)
+    {
+      half2 *C = reinterpret_cast<half2 *>(smem);
+
+      const int idx_strided = warp_row * 16 + wrm.quad_row * 8 + wrm.quad_hilo * 4 + (wrm.quad_thread % 2);
+      const int idx_contiguous = warp_col * 8 + wrm.quad_col * 4 + (wrm.quad_thread / 2);
+
+      int thread_offset_c = idx_strided * stride + idx_contiguous;
+      C[thread_offset_c] = __floats2half2_rn(reg[0], reg[1]);
+
+      thread_offset_c = (idx_strided + 2) * stride + idx_contiguous;
+      C[thread_offset_c] = __floats2half2_rn(reg[2], reg[3]);
+
+      thread_offset_c = idx_strided * stride + (idx_contiguous + 2);
+      C[thread_offset_c] = __floats2half2_rn(reg[4], reg[5]);
+
+      thread_offset_c = (idx_strided + 2) * stride + (idx_contiguous + 2);
+      C[thread_offset_c] = __floats2half2_rn(reg[6], reg[7]);
+    }
+  };
+
+  template <int BlockDimX, int Ls, int M, int N, int M_PAD, int N_PAD, bool reload, class T>
+  __device__ inline void mma_sync_gemm(T op_a[], half *sm_a, half *sm_b, half *sm_c, const WarpRegisterMapping &wrm)
```

Comment:
Do we need a `__CUDA_ARCH__` check within this routine? i.e., will it barf when compiled against `sm60`?

Reply:
Nope.  This is fine since the file is included conditionally on `__CUDA_ARCH__`, Jenkins compilation tests pass, and I've tested this by hand already.  All is fine.